### ISSUE TITLE
OpenSSL 1.1.x : unable to connect via Camellia encryption

### ISSUE THREAD

**Issue by wolfbeast** [openssl/openssl#11061]
OpenSSL 1.1.0/1.1.1 versions are unable to connect to servers that restrict encryption to Camellia suites (disabling AES, 3DES and RCx). Previous versions of OpenSSL could connect.

Camellia is an AES-comparable cipher, and considered strong and modern. As such this inability to connect seems to be a bug. Was this cipher disabled by mistake?

**Comment by kroeckx** [comment]
It's not in the default list of ciphers, but you can enable it.

Note that CAMELLIA doesn't have any AEAD ciphers and is not
supported in TLS 1.3.

**Comment by wolfbeast** [comment]
I'm aware it's not supported in TLS 1.3 which only has like 2 suites.
AEAD should be possible with Camellia too (e.g. GCM), why hasn't that been done?
Does enabling it require building from source, or can it be done through run-time config?

**Comment by kroeckx** [comment]
It can be enabled at runtime using the cipher string. See for
instance the cipher manpage, and the -cipher option for s_client
and s_server.

I don't know if there are any RFCs that define Camellia with GCM.
If such an RFC exists, it should be easy to add it to OpenSSL,
but we don't see any demand for it.

**Comment by wolfbeast** [comment]
RFC 6367 section 2.2 lists full ranges of GCM for Camellia.

https://tools.ietf.org/html/rfc6367#section-2.2

> but we don't see any demand for it.
The demand exists. This was discussed before when it was added to Pale Moon. It would be re-added to the client if a major lib like OpenSSL and applications building on it would actually implement this RFC completely.

**Comment by kroeckx** [comment]
I really don't see the need for it. If you want an alternative to
AES, there is Chacha20. Camellia is going to be slow compared to
both of them.

Our Camellia implementation is probably also not constant time.

**Comment by kroeckx** [comment]
So just to clarify, even when the GCM variants get added, it's
very unlikely that we will enable them by default. And you're
always free to submit pull requests for it.

**Comment by wolfbeast** [comment]
> Camellia is going to be slow compared to both of them.

Not necessarily. It should be able to use the AES-NI instructions in modern CPUs for the S-Box in its implementation. In addition it was designed to be a low-resource cipher.

> So just to clarify, even when the GCM variants get added, it's very unlikely that we will enable them by default.

As long as it's available as an option, the rest can be determined by run-time configuration, so that wouldn't upset your status quo. I'm still of the opinion that it -should- be part of the default set, but you're free to disagree.
As for submitting a pull request, I'm afraid I'm not well-versed enough in either the language C or the specific code structure in use in the library to make a meaningful code commit. :/

**Comment by mattcaswell** [comment]
> The demand exists. This was discussed before when it was added to Pale Moon. It would be re-added to the client if a major lib like OpenSSL and applications building on it would actually implement this RFC completely.

I don't ever recall anyone asking for CAMELLIA GCM suites before. At least not in recent memory. Nonetheless, if someone feels strongly enough about it to submit a PR then it would be considered and I can't think of a reason why we wouldn't add it.

I agree with @kroeckx though that we would be unlikely to add these to default.

**Comment by wolfbeast** [comment]
As for a PR, apparently there was previously already a PR made for this but the submitter apparently got tired of having to maintain it without it actually being merged in at the time. I don't know how much this has bitrotted since then, but perhaps it can be uplifted to the current state of things? PR #374

**Comment by roytam1** [comment]
for reference: https://gist.github.com/h-yamamo/157804dc9011a33afb99a3324114ee23

**Comment by demonfoo** [comment]
Will this ever be added? It seems like there's not a lot of cipher diversity these days, so having another AEAD suite seems like a good idea.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 3 months otherwise.

**Comment by marek22k** [comment]
I came across this issue while searching for "openssl camellia-gcm". I think it's great to have an alternative to AES.
Specifically, I want to create a VPN tunnel with someone via Tinc. Tinc uses OpenSSL, and my peering partner doesn't want to use AES. That's why we want to use Camellia. It would be great if GCM were also supported for Camellia.

**Comment by paulidale** [comment]
ChaCHa20/Poly?

**Comment by marek22k** [comment]
Tinc 1.0 requires (recommends it and is strongly designed for it in terms of code) a CBC cipher. If I am not mistaken, ChaCha20 is not a CBC cipher, but a stream cipher.

https://github.com/gsliepen/tinc/issues/482

Furthermore, Tinc 1.0 separates authentication (HMAC) and encryption (cipher). Can ChaCha20 also be used on its own? (As far as I know, yes, even if it is rarely done).

**Comment by paulidale** [comment]
I don't this last comment is relevant to this discussion about including **Camellia-GCM**.  Nonetheless:

* GCM is not a CBC cipher.  GCM is an AEAD cipher based on CTR mode.  Being AEAD, it includes authentication.
* Camellia using CBC mode **is** available.
* ChaCha20 can be used standalone.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity for more than 2 months.
